File: code/c/primenumbers.c
#include <stdio.h>

// Function to generate prime numbers up to nmax
void generate_primes(int nmax) {
    int x[5] = {2, 3, 5, 7, 11}; // Initial values
    int n = 1; // Counter of the current nth prime

    // Just print the preset values for n < 5
    if (nmax <= 5) {
        while (n <= nmax) {
            printf("%d, ", x[n-1]);
            n++;
        }
        printf("\n");
    } else {
        // Print initial values
        while (n <= 5) {
            printf("%d, ", x[n-1]);
            n++;
        }
        // Apply the recurrence relationship otherwise
        while (n <= nmax) {
            // Apply the formula
            int xn = x[4] + x[3] - x[2];
            int is_prime = 1; // Flag to check if xn is prime

            // Check if xn is prime
            for (int i = 0; i < 5; i++) {
                if (xn % x[i] == 0) {
                    is_prime = 0;
                    break;
                }
            }

            if (is_prime) {
                printf("%d, ", xn);
                // Update values
                for (int i = 0; i < 4; i++) {
                    x[i] = x[i+1];
                }
                x[4] = xn;
                n++;
            } else {
                // Update values
                for (int i = 0; i < 4; i++) {
                    x[i] = x[i+1];
                }
                x[4] = xn;
                n++;
            }
        }
        printf("\n");
    }
}

int main() {
    printf("Welcome to Dewan Mukto's Prime Number Generator!\n");
    int nmax;
    printf("Upto which prime to generate?: ");
    scanf("%d", &nmax);
    generate_primes(nmax);
    return 0;
}

File: code/python/PrimeNumbers.py
# Prime number generator
# @author Dewan Mukto
# @version 0.5.2.2024
def generate_primes(nmax):
    x = [2, 3, 5, 7, 11] # initial values
    n = 1 # counter of the current nth prime

    # just spit out the preset values for n < 5
    if nmax <= 5:
        while n <= nmax:
            print(x[n-1], end=", ")
            n += 1
        print()
    # apply the "recurrence relationship" otherwise
    else:
        while n <= 5:
            print(x[n-1], end=", ")
            n += 1
        while n <= nmax:
            # apply the formula
            xn = x[-1] + x[-2] - x[-3]
            if not any(xn % xi == 0 for xi in x):
                # if the Xn is not divisible by any of its previous "values"
                # (see Note on my website https://dewanmukto.github.io/research/mathematics/discovery/2024/05/02/prime-number-formula)
                print(xn, end=", ")
                x.append(xn)
                n += 1
            else:
                x.append(xn)
                n += 1
        print()
# The "recurrence relationship": Xn = Xn-1 + Xn-2 - Xn-3
print("Welcome to Dewan Mukto's Prime Number Generator!")
nmax = int(input("Upto which prime to generate?: "))
generate_primes(nmax)
